tag_manager/content_security_policy.rst

```diff
@@ -0,0 +1,156 @@
+            function stgCreateCookie(a,b,c){var d="";if(c){var e=new Date;e.setTime(e.getTime()+24*c*60*60*1e3),d=";expires="+e.toUTCString()}document.cookie=a+"="+b+d+"; path=/"}
+            var isStgDebug=(window.location.href.match("stg_debug")||document.cookie.match("stg_debug"))&&!window.location.href.match("stg_disable_debug");
+            stgCreateCookie("stg_debug",isStgDebug?1:"",isStgDebug?14:-1);
+            var qP=[];dataLayerName!=="dataLayer"&&qP.push("data_layer_name="+dataLayerName),isStgDebug&&qP.push("stg_debug");
+            var qPString=qP.length>0?("?"+qP.join("&")):"";
+            document.write('<script src="//client.containers.piwik.pro/'+id+'.sync.js' + qPString + '" nonce="INSERT_VALID_NONCE_VALUE"></' + 'script>');
+            })(window, document, 'dataLayer', 'feacd61d-0232-40a1-96c3-7e469f7bfa7f');
+        </script>
+
+.. note::
+    All that is needed for Tag Manager to work is to replace **INSERT_VALID_NONCE_VALUE** with generated nonce value. It should be done twice for both asynchronous and synchronous snippet.
+
+
+Adjusting tags to work with Content Security Policy
+---------------
+
+-   **asynchronous tags** - in most cases there should not be any change required to make asynchronous tags work. Tag Manager will automatically insert nonce attribute to all fired tags. Only exceptions is when Your tag adds other scripts/styles on page by itself - in such case, You should add nonce attribute manually.
+-   **synchronous tags** - since synchronous tags have to fire before whole page is loaded, following procedure is recommended:
+
+
+    1.  Create new variable with value of nonce parameter. It is not required to create nonce variable in admin panel. Just pushing it on dataLayer before script is executed is enough.
+
+        .. code-block:: javascript
+
+            window.dataLayer.push({
+                nonce: INSERT_VALID_NONCE_VALUE
+            });
+
+
+    2.  Use created variable as value for nonce attribute like follows:
+
+        .. code-block:: html
+
+            <script nonce="{{ nonce }}">
+                console.log("I'm synchronous tag!");
+                document.write('<p id="synchronous-tag">I was inserted by synchronous tag</p>');
+            </script>
+
+.. note::
+    Finally, not all 3rd party tools that are available as build-in templates are adjusted to work with Content Security Policy. This includes e.g. Google Analytics. In such cases, please refer to documentation of each respective tool (e.g. https://developers.google.com/web/fundamentals/security/csp).
+
+
+Tag Manager debugger
+--------------
+
+To load all necessary assets from Tag Manager debugger you need to define source with ``img-src``, ``font-src`` and ``style-src`` directives:
+
+.. code-block:: javascript
+
+	img-src <your-sources> client.containers.piwik.pro;
+	font-src <your-sources> client.containers.piwik.pro;
+	style-src <your-sources> 'nonce-nceIOfn39fn3e9h3sd';
```

Comment:
```suggestion
	style-src <your-sources> client.containers.piwik.pro;
```

Reply:
Good point

Reply:
However with nonce it still works :)

Reply:
And I probably need to stay with nonce solution because it support solutions where we have communication sources like:
- client.containers.piwik.pro
- client.piwik.pro

Setups like our dev environments (dev01 and dev02) where we are using one domain for everything (Piwik tracker, consent collecting, fetching containers by /containers) could work with: 
```style-src 'self' pptm-dev02.piwik.pro;```
but in staging, clouds and similar envs it does not work even when we configure it like this:
```style-src 'self' staging-qa.piwik.pro staging-qa.containers.piwik.pro;```

Reply:
I have to disagree here,
debugger doesn't have inline styles nor can it use nonce
even on staging, which you mention, debugger files are served from `staging-qa.containers.piwik.pro`
and setting `default-src 'self' staging-qa.piwik.pro staging-qa.containers.piwik.pro 'nonce-<cut>'` is enough for debugger to work so mentioning in this section rule with nonce is not a proper information

Reply:
I verified it again and you are right. I mixed up two cases: debugger assets and consent assets. It seems that nonce part is necessary for consent assets. I need to verify it one more time.